### ISSUE TITLE
Proposal for new way of handling text metadata

### ISSUE THREAD

**Issue by rmunn** [sillsdev/web-languageforge#1147]
**Is your feature request related to a problem? Please describe.**

Currently there's a mismatch between the FLEx data model and the LF data model for text. In FLEx, text is stored with a class called TsString, which contains text *and* metadata such as markup. The markup can be styles (bold, italics, etc.) but it's most often used for tagging the language that one span of text is in, which determines the font used to render that span. E.g., in a Notes field there might be a combination of English and Greek, like "The word λόγος has certain connotations..."

In Language Forge's data model, there is only one string for each writing system, and no concept of metadata attached to that string. This data model mismatch has been handled in the Send/Receive system by adding `<span>` tags, like `<span lang="grc">λόγος</span>`, but there are several issues with that, because the Language Forge UI doesn't handle metadata at all and so that metadata risks being lost if users edit the text field. (If they delete the word λόγος, then add it back, there's no way for them to know that the `<span>` tag has been deleted, and so when the next Send/Receive is done, λόγος will end up tagged as English text in FLEx rather than Greek text, meaning it will probably use the wrong font and the Greek expert on the project will be annoyed that it looks wrong.)

**Describe the solution you'd like**

I propose storing metadata alongside the content of text fields, rather than in a `<span>` that gets inserted into the text. Currently, any text field has the JSON structure `{ value: "contents" }`, i.e. it's stored as an object rather than as a simple string. I propose changing the LexValue class in PHP to have a second, optional, property `metadata`. I haven't yet decided what the contents of `metadata` will be, but they will probably be a structure similar to how TsString is implemented in FLEx: a series of spans of text, defined with start and end index positions, and the metadata about that span of text (what style does it have, what language is it in, and so on). The start and end index positions would be Python-style, where the start is (index of first character) and the end is (index of last character **plus one**). This has several useful properties: first, then span length is just (end - start), and second, consecutive spans have the start of span 2 equal to the end of span 1.

**Describe alternatives you've considered**

An alternative approach is to store the metadata at the field level rather than at the text level, using a field naming scheme like "leading underscore before a field name means metadata for that field". (We would also then declare that leading underscores are reserved for metadata, and no field name is allowed to have a leading underscore. I don't believe that would be a problem in our current data.)

This would allow us to eventually "flatten" the LexValue from being `{ value: "contents" }` to just being the simple string "contents", which would save a lot of space in our Mongo tables, and make dealing with our Mongo database in other technologies (like Node.js) quite a bit simpler. If we add a `metadata: { some structure }` field to LexValue we'll have a lot more trouble in the future when we decide to rewrite our backend.

The only drawback to this proposal is that the field name `_id` is used by MongoDB to store the ObjectId of any MongoDB document, so if we ever created a field named "id" we would not be able to store metadata for that field. If a user had a custom field named "id", though, that would not be a problem, because we store all custom fields under a "customFields" subdocument, so there would be no name conflict with `_id`.

I somewhat prefer this solution because it would allow us to move away from the `{ value: "contents" }` structure entirely. However, it would mean that the UI controls would have to be passed a field value *and* an optional metadata object if they were going to apply the metadata. Which might be tricky.

**Additional context**

This is a big proposal, and is not fully fleshed out yet. (For example, the contents of the `metadata` property are not yet defined.) There would also probably be implications for the UI if we have metadata using text spans: see the "Unanswered questions" section for more details.

**Unanswered questions**
* Does the UI need to keep track of span indices, and adjust them when characters are deleted?
* What do the indices represent, Unicode characters or codepoints?
    * Emoji and other characters on the "astral plane" (where one Unicode character is represented by *two* UTF-16 codepoints) would need to be carefully tested so that the spans indices would not end up being wrong when non-BMP characters are in that field.
    * We would also need to look at how the span indices for non-BMP characters are handled in FLEx, which is written in C# and therefore deals with the same UTF-16 encoding issues that Javascript has.
    * It would probably be best to mirror whatever FLEx does with indices — characters or codepoints — so that the Send/Receive process doesn't have to make index adjustments to metadata.

**Comment by alex-larkin** [comment]
I agree that the ease with which metadata can be silently lost in a string on Language Forge is concerning.

**Comment by megahirt** [comment]
@rmunn thank you for writing up this proposal.  There is a lot of good stuff in here for further discussion.  I think we could implement your proposal in two distinct phases.  We could silently start to populate the metadata into the model before we make use of it anywhere.  Phase 2 would attempt to make use of it.

I want to remind @rmunn that current LF code protects against data loss/corruption by disabling any field which contains a `<span>` tag.   See https://github.com/sillsdev/web-languageforge/search?q=modelContainsSpan

@alex-larkin there is no issue with losing metadata silently as Robin said - he wasn't thinking about the code I reference above which protects against that.  There is a longer story behind this situation, and the long-short is that we tried quite hard to implement a WYSIWYG editor for fields which contained markup/language strings, but we couldn't get it to work, so I made the decision to simply not allow the user to change data that contains `<span>`.  At least the data is safe in this case.
